t/semaphore.t

```diff
@@ -0,0 +1,1431 @@
+our $HttpConfig = <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+_EOC_
+
+our $HttpConfigInitByLua= <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+    init_by_lua '
+            require "resty.core.semaphore"
+            local sem  = ngx.semaphore.new(0)
+            local ok,err=sem:wait(1)
+            ngx.semaphore.err = err
+    ';
+_EOC_
+
+our $HttpConfigIntWorkerBy = <<_EOC_;
+    lua_package_path "$pwd/lib/?.lua;";
+    init_worker_by_lua '
+            require "resty.core.semaphore"
+            local sem  = ngx.semaphore.new(0)
+            local ok,err=sem:wait(1)
+            ngx.semaphore.test = err
+    ';
+
+_EOC_
+
+run_tests();
+
+__DATA__
+
+
+=== TEST 1: basic semaphore wait post
+--- http_config eval: $::HttpConfig
+--- config
+    location = /test {
+        access_log off;
+        content_by_lua '
+        local res1, res2 = ngx.location.capture_multi{
+          { "/sub_sem_wait"},
+          { "/sub_sem_post"},
+        }
+        ngx.say(res1.status)
+        ngx.say(res1.body)
+        ngx.say(res2.status)
+        ngx.say(res2.body)
+        ';
+    }
+     location /sub_sem_wait {
+        proxy_pass $scheme://127.0.0.1:$server_port/sem_wait;
+    }
+
+    location /sub_sem_post {
+        proxy_pass $scheme://127.0.0.1:$server_port/sem_post;
+    }
+
+    location /sem_wait {
+        content_by_lua '
+            require "resty.core.semaphore"
+            if not ngx.semaphore.test then
+                local sem,err = ngx.semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                ngx.semaphore.test = sem
+            end
+            local sem  = ngx.semaphore.test
+            local ok,err = sem:wait(10)
+            if ok then
+                ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.exit(500)
+            end
+        ';
+    }
+
+    location /sem_post {
+        content_by_lua '
+            require "resty.core.semaphore"
+            if not ngx.semaphore.test then
+                local sem,err = ngx.semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                ngx.semaphore.test = sem
+            end
+            local sem  = ngx.semaphore.test
+            local ok,err = sem:post()
+            if ok then
+                ngx.print("post")
+                ngx.exit(200)
+            else
+                ngx.exit(500)
+            end
+
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- no_error_log
+[error]
+
+
+
+=== TEST 2: basic semaphore wait post
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+        content_by_lua '
+            require "resty.core.semaphore"
+            local sem  = ngx.semaphore.new(0)
+            local ok,err=sem:wait(1)
```

Comment:
Style: please be careful about spaces. We need exactly one space before and after infix operators :)